pkg/mutation/assign_mutator.go

```diff
@@ -131,3 +135,80 @@ func applyToToBindings(applyTos []mutationsv1alpha1.ApplyTo) []SchemaBinding {
 	}
 	return res
 }
+
+// IsValidAssign returns an error if the given assign object is not
+// semantically valid
+func IsValidAssign(assign *mutationsv1alpha1.Assign) error {
+	path, err := parser.Parse(assign.Spec.Location)
+	if err != nil {
+		return errors.Wrapf(err, "Invalid location format")
+	}
+
+	if hasMetadataRoot(path) {
+		return fmt.Errorf("Assign can't change metadata")
+	}
+
+	err = validatePath(path)
+	if err != nil {
+		return err
+	}
+
+	toAssign := make(map[string]interface{})
+	err = json.Unmarshal([]byte(assign.Spec.Parameters.Assign.Raw), &toAssign)
+	if err != nil {
+		return errors.Wrapf(err, "Invalid format for parameters.assign")
+	}
+
+	value, ok := toAssign["value"]
+	if !ok {
+		return fmt.Errorf("spec.parameters.assign must have a value field")
+	}
+
+	err = validateObjectAssignedToList(path, value)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func hasMetadataRoot(path *parser.Path) bool {
+	if len(path.Nodes) == 0 {
+		return false
+	}
+
+	if reflect.DeepEqual(path.Nodes[0], &parser.Object{Reference: "metadata"}) {
+		return true
+	}
+	return false
+}
+
+func validatePath(p *parser.Path) error {
+	if len(p.Nodes) == 0 || p.Nodes == nil {
+		return fmt.Errorf("Empty path")
+	}
+	if len(p.Nodes) < 2 {
+		return nil
+	}
+	lastNode := p.Nodes[len(p.Nodes)-1]
+	secondLastNode := p.Nodes[len(p.Nodes)-2]
+	// the format spec.containers[name:foo].bar is not allowed
+	if secondLastNode.Type() == parser.ListNode && lastNode.Type() == parser.ObjectNode {
+		return fmt.Errorf("Invalid path format: child a list can't be an object")
+
+	}
+	return nil
+}
+
+func validateObjectAssignedToList(p *parser.Path, value interface{}) error {
+	if len(p.Nodes) == 0 || p.Nodes == nil {
+		return fmt.Errorf("Empty path")
+	}
+	if p.Nodes[len(p.Nodes)-1].Type() != parser.ListNode {
+		return nil
+	}
+	_, ok := value.(map[string]interface{})
```

Comment:
+1 this is awesome!

The last step here is to validate that:

`valueMap[listNode.KeyField] == listNode.KeyValue`

(the field must both be present and equal to the expected value)